internal/gcs/bridge.go

```diff
@@ -0,0 +1,356 @@
+	bridgeFailureTimeout = time.Minute * 5
+)
+
+type notifyFunc func(*containerNotification) error
+
+// newBridge returns a bridge on the given connection. It calls notify when a
+// notification message arrives from the guest. It logs transport errors and
+// traces using the given logrus entry.
+func newBridge(conn io.ReadWriteCloser, notify notifyFunc, log *logrus.Entry) *bridge {
+	return &bridge{
+		conn:    conn,
+		rpcs:    make(map[int64]*rpc),
+		rpcCh:   make(chan *rpc),
+		notify:  notify,
+		log:     log,
+		Timeout: bridgeFailureTimeout,
+	}
+}
+
+// Start begins the bridge send and receive goroutines.
+func (brdg *bridge) Start() {
+	go brdg.recvLoopRoutine()
+	go brdg.sendLoopRoutine()
+}
+
+// Kill terminates the bridge, closing the connection and causing all new and
+// existing RPCs to fail.
+func (brdg *bridge) Kill() {
+	brdg.mu.Lock()
+	if brdg.closed {
+		brdg.mu.Unlock()
+		return
+	}
+	brdg.closed = true
+	brdg.mu.Unlock()
+	brdg.log.Debug("bridge terminating")
+	brdg.conn.Close()
+}
+
+// Close closes the bridge. It is undefined to call RPC or AsyncRPC after
+// calling Close.
+func (brdg *bridge) Close() error {
+	brdg.Kill()
+	close(brdg.rpcCh)
+	return nil
+}
+
+// AsyncRPC sends an RPC request to the guest but does not wait for a response.
+// If the message cannot be sent before the context is done, then an error is
+// returned.
+func (brdg *bridge) AsyncRPC(ctx context.Context, proc rpcProc, req requestMessage, resp responseMessage) (*rpc, error) {
+	call := &rpc{
+		ch:   make(chan struct{}),
+		proc: proc,
+		req:  req,
+		resp: resp,
+	}
+	if err := ctx.Err(); err != nil {
+		return nil, err
+	}
+	// Send the request.
+	select {
+	case brdg.rpcCh <- call:
+		return call, nil
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	}
+}
+
+func (call *rpc) complete(err error) {
+	call.err = err
+	close(call.ch)
+}
+
+// Err returns the RPC's result. This may be a transport error or an error from
+// the message response.
+func (call *rpc) Err() error {
+	if call.err != nil {
+		return call.err
+	}
+	return call.resp.Err()
+}
+
+// Done returns whether the RPC has completed.
+func (call *rpc) Done() bool {
+	select {
+	case <-call.ch:
+		return true
+	default:
+		return false
+	}
+}
+
+// RPC issues a synchronous RPC request. Returns immediately if the context
+// becomes done and the message is not sent.
+//
+// If allowCancel is set and the context becomes done, returns an error without
+// waiting for a response. Avoid this on messages that are not idempotent or
+// otherwise safe to ignore the response of.
+func (brdg *bridge) RPC(ctx context.Context, proc rpcProc, req requestMessage, resp responseMessage, allowCancel bool) error {
+	call, err := brdg.AsyncRPC(ctx, proc, req, resp)
+	if err != nil {
+		return err
+	}
+	var ctxDone <-chan struct{}
+	if allowCancel {
+		// This message can be safely cancelled by ignoring the response.
+		ctxDone = ctx.Done()
+	}
+	t := time.NewTimer(brdg.Timeout)
```

Comment:
I can't remember where we landed on this but it looks like you decided against `context.WithDeadline(parent, timeout)` and instead are using the timer. Do you not want to handle parent cancellations "pre send"?

Reply:
AsyncRPC will handle the parent timeouts pre send.